### DISCUSSION TITLE
Why is this text shadow not working?

### DISCUSSION THREAD

**Discussion by SentientBeingOnAQuest** [ImageMagick/ImageMagick#4614] [Help] [answered]
Sorry for all the noob questions I'm asking, I'm really struggling here.
I'm trying to make this text shadow:

![font_denser_soft_outline](https://user-images.githubusercontent.com/95970582/146671787-4dbb6023-104c-42d5-bfe3-c93005918ae3.jpg)

But what I'm ending up with instead is:

![result_maxresdefault (1)](https://user-images.githubusercontent.com/95970582/146671801-e6e86b69-3f24-4425-bdcc-86806e0f6ec7.jpg)

Here's the command I'm using:

```
for %F in (*.jpg) do magick "%F" ^
-resize 572x ^
+write mpr:img +delete ^
( mpr:img -blur 10x10 -resize x592 -gravity center -crop 592x592+0+0 +repage ) ^
( mpr:img -bordercolor #DB9D00 -border 1x1 -bordercolor #FDF9DC -border 1x1 -bordercolor #DB9D00 -border 1x1 +repage ) ^
-gravity center -compose over -composite ^
-bordercolor #DB9D00 -border 2x2 -bordercolor #FDF9DC -border 2x2 -bordercolor #DB9D00 -border 2x2 ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -blur 0x8 ) -composite ^
-background none -fill white -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -composite ^
"result_%F"
```

How to fix this command? (IM7Win10)

**Comment by snibgo** [comment]
I suppose your "Anthony" example used "-shadow" command. Instead, you are doing this with blurred text, and unblurred text composited over that.

Your first text is filled with white, with a thick (8) blue stroke. Your second text is filled with white, and you haven't changed the stroke colour or width, so that will also be a thick blue stroke.

I suggest "-stroke none" before the second "caption:".

**Reply by SentientBeingOnAQuest** [reply]
I'm actually using the same principle as the Anthony example. (adapted for Win10 obviously) from imagemagick.org:
```
convert -size 320x100 xc:lightblue -font Candice -pointsize 72 \            
-stroke black -strokewidth 8 -annotate +25+65 'Anthony' -blur 0x8 \           
 -fill white   -stroke none   -annotate +25+65 'Anthony' \           
 font_denser_soft_outline.jpg

```
The -shadow command doesn't accurately outline every region of the text, it just produces a shadow roughly behind the text. 
I've tried this command with -shadow and it doesn't look good. 

Also there is no stroke in the overlay text. Just in the Shadow text. The idea is that they overlay one and other and only the stroke shows from behind the foreground text: 

Shadow:
```
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -blur 0x8 ) -composite
```

Overlay text:
```
-background none -fill white -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -composite

```

**Reply by fmw42** [reply]
That is not a Windows command syntax.  You need to change \ to ^ for Windows.

**Reply by SentientBeingOnAQuest** [reply]
I know, that's why I said: (adapted for Win10 obviously)
In the original post is the command I'm using.

```
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -blur 0x8 ) -composite ^
-background none -fill white -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -composite ^
```

The principle should work the same as with the Anthony example, yet it doesn't work.

**Reply by fmw42** [reply]
Please post your full command line exactly as you use it and your input and output images. We cannot critique a command unless we see the full command. Please use end of line ^ to format the the command and then post using the <> button to keep it formatted exactly as you use it to help us read your command more easily.

**Reply by SentientBeingOnAQuest** [reply]
All of that is in the original post. The input image is the same except just without the text. Please take a look at my original post with the question: https://github.com/ImageMagick/ImageMagick/discussions/4614#discussion-3755263

**Reply by fmw42** [reply]
Post your input image before adding the text so we can test with it. The command in the original post does not look like Anthony's. Perhaps you should try his command first and make sure you can reproduce it as well as then using your approach without all the excess code for other things and for looping. Start simple and then expand.

**Reply by SentientBeingOnAQuest** [reply]
Yes, sorry I just realised the input image is relevant since I have a lot of other stuff going on in that command.

Input image: 

![maxresdefault (1)](https://user-images.githubusercontent.com/95970582/146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg)

**Reply by fmw42** [reply]
Try as short a command as you can that just tries to add the blurred text. We can critique that and perhaps get you over the hurdle

**Reply by SentientBeingOnAQuest** [reply]
The shortest command would be: 
```
for %F in (*.jpg) do magick "%F" ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -blur 0x8 ) -composite ^
-background none -fill white -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -composite ^
"result_%F"
```

Which produces:
![result_maxresdefault (1)](https://user-images.githubusercontent.com/95970582/146702676-9674c3c5-9926-4323-8ac5-73355697f4aa.jpg)

The rest of the command works perfectly (thanks to you guys 👍 ) and doesn't deal with the text part.

**Reply by fmw42** [reply]
Actually that is not the shortest. You have included code for looping. However, that is minor.  But what do you want it to look like or what is wrong with the above.

**Reply by SentientBeingOnAQuest** [reply]
I left the loop in since I thought it would be easier than having to edit the filename all the time. However without the for loop it would be:
```
magick input.jpg ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -blur 0x8 ) -composite ^
-background none -fill white -font San-Diego -pointsize 70 -size 550x caption:"MARLINGBAITS" -gravity North -geometry +0+15 -composite ^
result.jpg
```

There are two issues:
1- The shadow is being cut off at the left side really sharply which doesn't look natural.
2- The overlaid text in white isn't rendering properly. It should be clearly readable as a regular text in the San-Diego font.
Like this, but with the blue shadow:

![result](https://user-images.githubusercontent.com/95970582/146705710-b27aef08-e799-411f-a9ed-38d5745d75a9.jpg)



What I'm trying to accomplish is this:

![font_denser_soft_outline](https://user-images.githubusercontent.com/95970582/146705354-de7193a7-d476-4fa5-8db3-9b044e36bf0a.jpg)

**Reply by fmw42** [reply]
The cutoff is due to the spread of the blur past the left edge of the caption. The only way to avoid that is to make the blur smaller or make -size larger. 

The rest of your script is almost fine except you have to reset the -gravity and -geometry after the first use so that it does not accumulate. The parentheses are not perfect. Also if you want make the blue lighter you can adjust the transparency.  See if this helps.  Also learn to use temporary images so that you can see what is happening along the way at each step. You can cut them out later. See my code below.  Also learn to use the <> button above the box so that you can keep your text formatted. Also put your ^ earlier in the line so that your lines are shorter and will stay on the page without scrolling.

Here is my version in Unix syntax:

```
magick 146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg \
\( -background none -fill blue -stroke blue -strokewidth 8 -font /library/fonts/Sandiego.ttf \
-pointsize 70 -size 550x caption:"MARLINGBAITS" +stroke -blur 0x8 \
-channel a -evaluate multiply 0.5 +channel +write t1.png \) \
-gravity North -geometry +0+15 -composite +geometry -gravity northwest +write t2.png \
\( -background none -fill white -stroke white -strokewidth 1 -font /library/fonts/Sandiego.ttf \
-pointsize 70 -size 550x caption:"MARLINGBAITS" +write t3.png \) \
-gravity North -geometry +0+15 -composite \
result.jpg

```

Reformat for Windows and look at the temp images t1, t2, t3

![result](https://user-images.githubusercontent.com/13458218/146706872-0bc57c73-ec55-447b-8436-e977248aadef.jpg)



Note that I did not try to fix it for the cut off blur.

**Reply by fmw42** [reply]
P.S. I do not think I found the same San-Diego font as you.

**Reply by fmw42** [reply]
Here is the same command with a different font and a bit more opacity:

```
magick 146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg \
\( -background none -fill blue -stroke blue -strokewidth 8 -font ubuntu \
-pointsize 70 -size 550x caption:"MARLINGBAITS" +stroke -blur 0x8 \
-channel a -evaluate multiply 0.75 +channel +write t1.png \) \
-gravity North -geometry +0+15 -composite +geometry -gravity northwest +write t2.png \
\( -background none -fill white -stroke white -strokewidth 1 -font ubuntu \
-pointsize 70 -size 550x caption:"MARLINGBAITS" +write t3.png \) \
-gravity North -geometry +0+15 -composite \
result.jpg

```

![result](https://user-images.githubusercontent.com/13458218/146707062-aa2a10c0-5d2b-4bcb-8da7-03b89400ed24.jpg)

**Reply by fmw42** [reply]
If you center the text in the caption, you will mitigate the sharp cutoff of the blur.


```
magick 146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg \
\( -background none -fill blue -stroke blue -strokewidth 8 -font ubuntu \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" +stroke -blur 0x8 \
-channel a -evaluate multiply 0.75 +channel +write t1.png \) \
-gravity North -geometry +0+15 -composite +geometry -gravity northwest +write t2.png \
\( -background none -fill white -stroke white -strokewidth 1 -font ubuntu \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" +write t3.png \) \
-gravity North -geometry +0+15 -composite \
result.jpg

```

![result](https://user-images.githubusercontent.com/13458218/146707201-e83527b6-c994-4f7c-a60b-ad972b138fd5.jpg)

**Reply by fmw42** [reply]
Here is full opacity on the text:

```
magick 146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg \
\( -background none -fill blue -stroke blue -strokewidth 8 -font ubuntu \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" +stroke -blur 0x8 \
-channel a -evaluate multiply 1 +channel +write t1.png \) \
-gravity North -geometry +0+15 -composite +geometry -gravity northwest +write t2.png \
\( -background none -fill white -stroke white -strokewidth 1 -font ubuntu \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" +write t3.png \) \
-gravity North -geometry +0+15 -composite \
result.jpg
```

![result](https://user-images.githubusercontent.com/13458218/146707286-731f13aa-eea4-4978-86c3-90690b08187b.jpg)

**Reply by SentientBeingOnAQuest** [reply]
Great stuff, this certainly seems to be working a lot better. I will try with my font. In the meantime I have done some experimenting with the -pointsize and the -size, but this does not seem to eliminate the cut edge. In fact it seems to be drawing a box around the text even if I leave -size away entirely:

![result_input](https://user-images.githubusercontent.com/95970582/146707499-360124ea-14bb-4a04-bc2a-5a086b0b6282.jpg)

So using your code adapted for Win10:
```
magick input.jpg ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego ^
-pointsize 70 -size 550x caption:"MARLINGBAITS" +stroke -blur 0x8 ^
-channel a -evaluate multiply 0.5 +channel +write t1.png ) ^
-gravity North -geometry +0+15 -composite +geometry -gravity northwest +write t2.png ^
( -background none -fill white -stroke white -strokewidth 1 -font San-Diego ^
-pointsize 70 -size 550x caption:"MARLINGBAITS" +write t3.png ) ^
-gravity North -geometry +0+15 -composite ^
result.jpg
```

Certainly looks a lot better:
![result](https://user-images.githubusercontent.com/95970582/146708208-d476092f-a9e6-4f56-9776-9c562478d182.jpg)


However I'm still getting the cutoff at the side.

**Reply by fmw42** [reply]
I have shortened the command to remove the +geometry, etc as I think the font was bad and so these are not needed. I also took out the temp files, now that it is working. 

```
magick 146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg \
\( -background none -fill blue -stroke blue -strokewidth 8 -font ubuntu \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" -blur 0x8 \) \
-gravity North -geometry +0+15 -composite  \
\( -background none -fill white -stroke white -strokewidth 1 -font ubuntu \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS"  \) \
-gravity North -geometry +0+15 -composite \
result.jpg

```

![result](https://user-images.githubusercontent.com/13458218/146707532-95bf7fb4-b5bb-46ea-a49a-021072743e1b.jpg)

**Reply by fmw42** [reply]
Center the text in the caption to mitigate the blur cutoff.

**Reply by SentientBeingOnAQuest** [reply]
ps there's a bit of an optical illusion going on, the text doesn't look straight due to the background image but it actually is.

**Reply by SentientBeingOnAQuest** [reply]
Shortened command edited for Win10:
```
magick input.jpg ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego ^
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" -blur 0x8 ) ^
-gravity North -geometry +0+15 -composite ^
( -background none -fill white -stroke white -strokewidth 1 -font San-Diego ^
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" ) ^
-gravity North -geometry +0+15 -composite ^
result.jpg
```

Result: 
![result](https://user-images.githubusercontent.com/95970582/146708763-55c51ee2-040c-4e52-a5e1-44fe084800fa.jpg)

I'd say that looks pretty darn good! 
So with the rest of the code it would be:

```
for %F in (*.jpg) do magick "%F" ^
-resize 572x ^
+write mpr:img +delete ^
( mpr:img -blur 10x10 -resize x592 -gravity center -crop 592x592+0+0 +repage ) ^
( mpr:img -bordercolor #DB9D00 -border 1x1 -bordercolor #FDF9DC -border 1x1 -bordercolor #DB9D00 -border 1x1 +repage ) ^
-gravity center -compose over -composite ^
-bordercolor #DB9D00 -border 2x2 -bordercolor #FDF9DC -border 2x2 -bordercolor #DB9D00 -border 2x2 ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego ^
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" -blur 0x8 ) ^
-gravity North -geometry +0+15 -composite ^
( -background none -fill white -stroke white -strokewidth 1 -font San-Diego ^
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" ) ^
-gravity North -geometry +0+15 -composite ^
"result_%F"
```

Produces: 

![result_maxresdefault (1)](https://user-images.githubusercontent.com/95970582/146708859-4d640fe7-41df-4ec3-afc4-414124607a5e.jpg)


Great! Thanks for your help again! If you repost that reply as an answer I can mark it as answered.

**Comment by fmw42** [comment, marked as answer]
Here is a different san diego font.

```
magick 146702178-abe4ae38-89df-41a6-9860-acb8f42fd2f3.jpg \
\( -background none -fill blue -stroke blue -strokewidth 8 -font /library/fonts/sandiego.ttf \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" -blur 0x8 \) \
-gravity North -geometry +0+15 -composite \
\( -background none -fill white -stroke white -strokewidth 1 -font /library/fonts/sandiego.ttf \
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" \) \
-gravity North -geometry +0+15 -composite \
result.jpg

```

![result](https://user-images.githubusercontent.com/13458218/146708865-c86956a1-4bba-47b6-8881-a513f2bca3a1.jpg)

**Reply by SentientBeingOnAQuest** [reply]
This works perfectly! Adapted for windows: 

magick input.jpg ^
( -background none -fill blue -stroke blue -strokewidth 8 -font San-Diego ^
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" -blur 0x8 ) ^
-gravity North -geometry +0+15 -composite ^
( -background none -fill white -stroke white -strokewidth 1 -font San-Diego ^
-pointsize 70 -size 550x -gravity center caption:"MARLINGBAITS" ) ^
-gravity North -geometry +0+15 -composite ^
result.jpg

Thank you very much!
